proposals/fips/fleet-server-fips.md

```diff
@@ -0,0 +1,62 @@
+
+Since we use Go 1.19, we only need to build fleet with `CGO_ENABLED=1` (already the case because of our sqlite3 dependency) and with `GOEXPERIMENT=boringcrypto`. This will automatically make Fleet use the BoringSSL cryptographic primitives instead of the stdlib `crypto` implementation (without requiring any code changes).
+
+See the [POC]https://github.com/fleetdm/fleet/compare/13288-poc-fleet-fips?expand=1()).
+
+> Source: https://kupczynski.info/posts/fips-golang/
+
+## Password hashing
+
+For password hashing Fleet currently uses the [bcrypt](https://en.wikipedia.org/wiki/Bcrypt) function.
+To be FIPS 140-2 compliant, Fleet will have to use [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2] instead (with SHA-256 as hashing primitive).
+
+> Source: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf
+
+## Summary of tasks
+
+Assumptions around the first customer using the FIPS build:
+- All MDM funtionality will be disabled/not-used.
+- The Fleet server won't be doing TLS termination. Thus we don't need to verify/test such feature.
+- This build will be used as part of a new deployment (not migrating an existing one). This is important because we have to change the password hashing algorithm thus to reduce complexity on the first iteration we don't need to worry about migrating from old to new password hashing.
+- The `fleetctl` command is outside the scope.
+
+Tasks:
+
+- Double check if any cryptographic operations in Fleet were missed in this analysis. // 1 pt.
+- Add a new target `fleet-fips` to the `Makefile` to build Fleet in FIPS mode. (Would set `GOEXPERIMENT=boringcrypto CGO_ENABLED=1`.) Smoke test the Fleet server when built this way. (See the [POC]https://github.com/fleetdm/fleet/compare/13288-poc-fleet-fips?expand=1()) // 1 pt.
+- Make changes in goreleaser yamls to create a new docker image fleetdm/fleet-fips // 2 pt.
+- Perform full QA of the Feet docker image FIPS build (by full QA we mean: we need to test ALL Fleet features). This should be performed by a QA engineer. // 5 pt.
+- Add tests to make sure the Fleet server TLS client is using the FIPS approved ciphers when connecting to TLS servers (SSO, users, vuln processing, webhooks). E.g. by using TLS servers that will fail the connection if the FIPS ciphers are not used. // 3 pt.
+- Replace bcrypt with PBKDF2 for password hashing (when compiling in FIPS mode) // 5 pt. (assumming we don't need to migrate from bcrypt)
+- Terminate the Fleet sever if MDM or TLS server is enabled when running the FIPS build (security measure). // 1 pt.
+- Check if `crypto/subtle` (used by the `/metrics` endpoint for HTTP basic auth) is implemented by BoringSSL. // 1 pt.
+- Loadtest the Fleet server with expected number of devices. // 5 pt. (depends on number of hosts)
+- Deploy the Fleet FIPS docker image and dependencies to a FIPS enabled AWS endpoint // 5 pt.
```

Comment:
Let's add the lines mentioned in the discussion:

- MDM work 
- Allow Migration (vs reinstall)
- Review fleetctl (timebox too)

Reply:
What about Fleetd?

Reply:
see above (if we implement FIPS on the server, and both fleetctl and fleetd need to authenticate to the server in order to transmit information, doesn't that necessarily mean we need to change fleetctl and fleetd to use different cryptographic algorithms?)

Reply:
Task added: [Add task for fleetd and osquery connection to a FIPS server](https://github.com/fleetdm/fleet/pull/13919/commits/955d8ea27c5c4638763039ceecdad0564be81904)